x/bank/keeper/send.go

```diff
@@ -189,6 +210,14 @@ func (k BaseSendKeeper) InputOutputCoins(ctx sdk.Context, inputs []types.Input,
 // SendCoins transfers amt coins from a sending account to a receiving account.
 // An error is returned upon failure.
 func (k BaseSendKeeper) SendCoins(ctx sdk.Context, fromAddr sdk.AccAddress, toAddr sdk.AccAddress, amt sdk.Coins) error {
+	if k.sendRestriction.Fn != nil {
+		var err error
+		toAddr, err = k.sendRestriction.Fn(ctx, fromAddr, toAddr, amt)
+		if err != nil {
+			return err
+		}
+	}
+
 	err := k.subUnlockedCoins(ctx, fromAddr, amt)
```

Comment:
You can go further and subsume the locking of coins as one type of restriction! (This will play nicely with the unification of locking, clawback, and staking contemplated in #14271.) This would change `subUnlockedCoins()` into just `subCoins()` for a nice symmetry with `addCoins()`. For `InputOutputCoins()`, you'd want to decrement the sender multiple times, once for each recipient, so the restriction check could observe the decremented balance.